public/app/features/correlations/Forms/TransformationsEditor.tsx

```diff
@@ -0,0 +1,281 @@
+  // may get bigger when the for is invalid
+  removeButton: css`
+    margin-top: 25px;
+  `,
+});
+
+export const TransformationsEditor = (props: Props) => {
+  const { control, formState, register, setValue, watch, getValues } = useFormContext();
+  const { readOnly } = props;
+  const [keptVals, setKeptVals] = useState<{ [key: string]: { expression: string; mapValue: string } }>({});
+
+  const styles = useStyles2(getStyles);
+
+  const transformOptions = getTransformOptions();
+
+  return (
+    <FieldArray name="config.transformations" control={control}>
+      {({ fields, append, remove }) => (
+        <>
+          <Stack direction="column" alignItems="flex-start">
+            <div className={styles.heading}>Transformations</div>
+            {fields.length === 0 && <div> No transformations defined.</div>}
+            {fields.length > 0 && (
+              <div>
+                {fields.map((field, index) => {
+                  return (
+                    <Stack direction="row" key={field.id} alignItems="top">
+                      <Field
+                        label={
+                          <Stack gap={0.5}>
+                            <Label>Type</Label>
+                            <Tooltip
+                              content={
+                                <div>
+                                  <p>The type of transformation that will be applied to the source data.</p>
+                                </div>
+                              }
+                            >
+                              <Icon name="info-circle" size="sm" />
+                            </Tooltip>
+                          </Stack>
+                        }
+                        invalid={!!formState.errors?.config?.transformations?.[index]?.type}
+                        error={formState.errors?.config?.transformations?.[index]?.type?.message}
+                        validationMessageHorizontalOverflow={true}
+                      >
+                        <InputControl
+                          render={({ field: { onChange, ref, ...field } }) => (
+                            <Select
+                              {...field}
+                              onChange={(value) => {
+                                if (!readOnly) {
+                                  const currentValues = getValues().config.transformations[index];
+                                  setKeptVals({
+                                    ...keptVals,
+                                    [index.toString()]: {
+                                      expression: currentValues.expression,
+                                      mapValue: currentValues.mapValue,
+                                    },
+                                  });
+                                  const newValueDetails = getSupportedTransTypeDetails(value.value);
+
+                                  if (newValueDetails.showExpression) {
```

Comment:
I don't think it will scale nicely if we add more transformations with different props to keep. Also the "kept state" should be cleared when removing rows, otherwise it looks like stale expression is filled after when removing then adding a new transformation.

I'd probably suggest to keep it simple for now, unless we consider it as critical for providing best UX (@radiorental)

Also: the dropdown is clickable in read-only mode. It looks a bit confusing, like the type could be edited. If it's needed to show the description of a transformation it'd probably work better as a tooltip.

Reply:
For future reference, the replication steps for the bug are
1. Create a new regex transformation, fill in all the fields
2. Change type to logfmt, see last 2 fields go blank
3. Delete that transformation
4. Create a new transformation. Select regex. See last two fields get filled.

Reply:
This should be fixed. Let me know if you find anything else.